rueidisprob/bloomfilter.go

```diff
@@ -0,0 +1,344 @@
+package rueidisprob
+
+import (
+	"context"
+	"errors"
+	"math"
+	"strconv"
+
+	"github.com/redis/rueidis"
+)
+
+const (
+	// NOTE: https://redis.io/docs/data-types/bitmaps/
+	maxSize = 1 << 32
+)
+
+const (
+	addMultiScript = `
+local hashIterations = tonumber(ARGV[1])
+local numElements = tonumber(#ARGV) - 1
+local filterKey = KEYS[1]
+local counterKey = KEYS[2]
+
+local bitfieldArgs = { filterKey }
+for i=2, numElements+1 do
+    table.insert(bitfieldArgs, 'SET')
+    table.insert(bitfieldArgs, 'u1')
+    table.insert(bitfieldArgs, ARGV[i])
+    table.insert(bitfieldArgs, '1')
+end
+
+local setBits = redis.call('BITFIELD', unpack(bitfieldArgs))
+
+local counter = 0
+local oneBits = 0
+for i=1, #setBits do
+	if setBits[i] == 1 then
+		oneBits = oneBits + 1
+	end
+
+	if i % hashIterations == 0 then
+		if oneBits ~= hashIterations then
+			counter = counter + 1
+		end
+
+		oneBits = 0
+	end
+end
+
+return redis.call('INCRBY', counterKey, counter)
+`
+
+	existsMultiScript = `
+local numElements = tonumber(#ARGV)
+local filterKey = KEYS[1]
+local result = {}
+
+local bitfieldArgs = { filterKey }
+for i=1, numElements do
+	local index = tonumber(ARGV[i])
+
+	table.insert(bitfieldArgs, 'GET')
+	table.insert(bitfieldArgs, 'u1')
+	table.insert(bitfieldArgs, index)
+end
+
+return redis.call('BITFIELD', unpack(bitfieldArgs))
```

Comment:
I am curious about whether would it be better if we calculate key existences here and return a bool array only instead of returning every bits? I am not sure which one will be better.

Reply:
there is trade-off. if we calculate on redis, we should use more CPU, but it can reduce bandwidth per request. In this situation, I prefer use less CPU. because bit array size is small event if keys and hash iteration size is large. but if you think bandwidth is more valuable than CPU, i will change it.

Reply:
We can leave this for future evaluation. I may be wrong, but my guess is that the calculation on redis is cheap because it just loops through the return of BITFIELD and sums up bits for every hash iteration. If we do that, we can save the bandwidth by hash iteration times. This requires more evaluation for verification.

I also have another idea: Can we use a normal GET to fetch the whole bitmap and do the calculation on the client side? By that, we can leverage client-side caching.

Reply:
[7491a90](https://github.com/redis/rueidis/pull/493/commits/7491a90298299bc8f7903b0f6369ed8b54267fb3)

OK. I change it.

Still, I don't think it is good to depend on Lua script heavily. scaling out servers is easy, but scaling out redis is not easy. 
I found another implementation of bloom filter using redis: https://github.com/igrigorik/bloomfilter-rb